software/contrib/quantizer.py

```diff
@@ -0,0 +1,488 @@
+        
+        # go through our scale and determine the nearest on-scale note
+        nearest_on_scale = 0
+        best_delta = 255
+        for note in range(len(self.scale)):
+            if self.scale[note]:
+                delta = abs(nearest_semitone - note)
+                if delta < best_delta:
+                    nearest_on_scale = note
+                    best_delta = delta
+            
+        self.current_note = nearest_on_scale
+        self.output_voltage = base_volts + (self.root + nearest_on_scale) * VOLTS_PER_SEMITONE + self.octave * VOLTS_PER_OCTAVE
+        
+    ## Read the input signal, quantize it, set outputs 1-5 accordingly
+    #
+    #  Called by the main loop in continuous mode or the rising clock handler
+    #  in triggered mode
+    def read_quantize_output(self):
+        self.input_voltage = ain.read_voltage(128)
+        self.quantize(self.input_voltage)
+        
+        cv1.voltage(self.output_voltage)
+        
+        for i in range(len(self.aux_outs)):
+            self.aux_outs[i].voltage(self.output_voltage + self.intervals[i] * VOLTS_PER_SEMITONE)
+    
+    ## Handler for the rising edge of the input clock
+    def on_trigger(self):
+        if self.mode == MODE_TRIGGERED:
+            self.read_quantize_output()
+            cv6.on()
+            
+    ## Handler for the falling edge of the input clock
+    def after_trigger(self):
+        if self.mode == MODE_TRIGGERED:
+            cv6.off()
+            
+    ## Handler for pressing button 1
+    #
+    #  Button 1 is used for the main interaction and is passed to
+    #  the current display for user interaction
+    def on_button1(self):
+        self.last_interaction_time = time.time()
+        self.active_screen.on_button1()
+        
+    ## Handler for pressing button 2
+    #
+    #  Button 2 is used to cycle between screens
+    def on_button2(self):
+        self.last_interaction_time = time.time()
+        
+        if self.active_screen == self.kb:
+            self.active_screen = self.menu
+        else:
+            self.active_screen = self.kb
+    
+    ## The main loop
+    #
+    #  Connects event handlers for clock-in and button presses
+    #  and runs the main loop
+    def main(self):
+        # connect the trigger handler here instead of the constructor
+        # otherwise it will start quantizing as soon as we instantiate the class
+        @din.handler
+        def on_rising_clock():
+            self.on_trigger()
+            
+        @din.handler_falling
+        def on_falling_clock():
+            self.after_trigger()
+            
+        @b1.handler
+        def on_b1_press():
+            self.on_button1()
+            
+        @b2.handler
+        def on_b2_press():
+            self.on_button2()
+        
+        while True:
+            # Update at 100Hz
+            CYCLE_RATE = 0.01
+            time.sleep(CYCLE_RATE)
```

Comment:
[discussion] Is it necessary to include this sleep? In my experience adding a sleep may cause the script to feel sluggish and not very responsive.

Reply:
I didn't notice it feeling sluggish in my testing. It's there mostly to ensure that the trigger signal generated by the continuous quantization mode is on for at least 10ms.

Reply:
Ah, I see, thanks for clarifying. Another strategy you can use here instead of sleep is to rely on the duration of time that has passed since the trigger went high. You could add code like this in the appropriate places:

```python
# Some code that starts a trigger
cv.on()
self._trigger_deadline = time.ticks_add(time.ticks_ms(), self.TRIGGER_DURATION)

...

# If trigger deadline has passed, then turn the trigger cv off
if time.ticks_diff(time.ticks_ms(), self._trigger_deadline ) > 0:
    cv.off()
```

Reply:
I'd considered that, but the problem with this approach is that we'd need to block continuous quantization until the trigger has elapsed. Otherwise the note could change twice within the duration of a single trigger's pulse width.  Blocking quantization until the previous trigger has cleared is basically the same thing as a sleep anyway, just more complicated to write.

Have you had a chance to test the script as-is with the sleep?  As I said, I didn't notice any sluggishness in my testing, but maybe I'm not as familiar with what "sluggish" feels like on the EuroPi.

Reply:
Ah, sounds like it's a "feature" then. Makes sense to me.

Reply:
> It's there mostly to ensure that the trigger signal generated by the continuous quantization mode is on for at least 10ms.

That sounds important, let's add a comment to that effect, so that we don't forget it when making changes in the future.

Would it make sense to move this sleep closer to the continuous quantization code? Maybe event into the `if` that turns cv6 on? or would this break something else?